README.md
# PSR-6 Cache Health Checker

[![Latest Version](https://img.shields.io/github/tag/oat-sa/lib-health-check-cache.svg?style=flat&label=release)](https://github.com/oat-sa/lib-health-check-cache/tags)
[![License GPL2](http://img.shields.io/badge/licence-GPL%202.0-blue.svg)](http://www.gnu.org/licenses/gpl-2.0.html)
[![Build Status](https://travis-ci.org/oat-sa/lib-health-check-cache.svg?branch=master)](https://travis-ci.org/oat-sa/lib-health-check-cache)
[![Coverage Status](https://coveralls.io/repos/github/oat-sa/lib-health-check-cache/badge.svg?branch=master)](https://coveralls.io/github/oat-sa/lib-health-check-cache?branch=master)
[![Packagist Downloads](http://img.shields.io/packagist/dt/oat-sa/lib-health-check-cache.svg)](https://packagist.org/packages/oat-sa/lib-health-check-cache)


> [PSR-6](https://www.php-fig.org/psr/psr-6/) cache health checker for the [Health checks PHP library](https://github.com/oat-sa/lib-health-check)

## Table of contents
- [Installation](#installation)
- [Usage](#usage)
- [Tests](#tests)

## Installation

```console
$ composer require oat-sa/lib-health-check-cache
```

## Usage

This library provides a [CacheChecker](src/CacheChecker.php) checker in charge to check if the provided cache pool is reachable.

```php
<?php

declare(strict_types=1);

use OAT\Library\HealthCheck\HealthChecker;
use OAT\Library\HealthCheckCache\CacheChecker;
use OAT\Library\HealthCheckCache\UuidCacheKeyGenerator;
use Psr\Cache\CacheItemPoolInterface;

$healthChecker = new HealthChecker();

/** @var CacheItemPoolInterface $cache */
$cacheChecker = new CacheChecker($cache);

$results = $healthChecker
    ->registerChecker($cacheChecker)
    ->performChecks();
```

> **_Note_**: The built-in [UuidCacheKeyGenerator](src/UuidCacheKeyGenerator.php) can accept a custom prefix. If not provided, a default prefix will be used.

> **_Note_**: If you need a custom logic to generate the cache key, create your own key generator by implementing [CacheKeyGeneratorInterface](src/CacheKeyGeneratorInterface.php)
> Make sure your generated key is unique enough.
## Tests

To run tests:
```console
$ vendor/bin/phpunit
```
> **_Note_**: see [phpunit.xml.dist](phpunit.xml.dist) for available test suites.


CHANGELOG.md
# CHANGELOG

## [1.0.0] - 2021-03-26

### Added
- `CacheChecker` class
- `RamseyCacheKeyGenerator` class


tests/Unit/CacheCheckerTest.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace OAT\Library\HealthCheckCache\Tests\Unit;

use Exception;
use OAT\Library\HealthCheckCache\CacheChecker;
use OAT\Library\HealthCheckCache\CacheKeyGeneratorInterface;
use PHPUnit\Framework\MockObject\MockObject;
use PHPUnit\Framework\TestCase;
use Psr\Cache\CacheException;
use Psr\Cache\CacheItemInterface;
use Psr\Cache\CacheItemPoolInterface;

class CacheCheckerTest extends TestCase
{
    private const CUSTOM_KEY = 'test-key';

    /** @var CacheChecker */
    private $subject;

    /** @var CacheItemPoolInterface|MockObject */
    private $cachePoolMock;

    /** @var CacheKeyGeneratorInterface|MockObject */
    private $keyGeneratorMock;

    protected function setUp(): void
    {
        $this->cachePoolMock = $this->createMock(CacheItemPoolInterface::class);
        $this->keyGeneratorMock = $this->createMock(CacheKeyGeneratorInterface::class);

        $this->subject = new CacheChecker($this->cachePoolMock, $this->keyGeneratorMock);
    }

    public function testGetIdentifier(): void
    {
        self::assertEquals('cache', $this->subject->getIdentifier());
    }

    public function testForSuccessCheckerResult(): void
    {
        $cacheItemMock = $this->createMock(CacheItemInterface::class);
        $cacheItemMock
            ->expects(self::once())
            ->method('set')
            ->with(self::CUSTOM_KEY);

        $cacheItemMock
            ->expects(self::once())
            ->method('expiresAfter')
            ->with(30);

        $cacheItemMock
            ->expects(self::once())
            ->method('isHit')
            ->willReturn(true);

        $cacheItemMock
            ->expects(self::once())
            ->method('get')
            ->willReturn(self::CUSTOM_KEY);

        $this->keyGeneratorMock
            ->expects(self::once())
            ->method('generate')
            ->willReturn(self::CUSTOM_KEY);

        $this->cachePoolMock
            ->expects(self::exactly(2))
            ->method('getItem')
            ->with(self::CUSTOM_KEY)
            ->willReturn($cacheItemMock);

        $this->cachePoolMock
            ->expects(self::once())
            ->method('save')
            ->with($cacheItemMock)
            ->willReturn(true);

        $this->cachePoolMock
            ->expects(self::once())
            ->method('deleteItem')
            ->with(self::CUSTOM_KEY)
            ->willReturn(true);

        $result = $this->subject->check();

        self::assertTrue($result->isSuccess());
        self::assertEquals(
            sprintf('Success on writing, reading and deleting cache item %s', self::CUSTOM_KEY),
            $result->getMessage()
        );
    }

    public function testForFailedCheckerResultWhenExceptionThrown(): void
    {
        $cacheException = new class('custom error') extends Exception implements CacheException{};

        $this->keyGeneratorMock
            ->expects(self::once())
            ->method('generate')
            ->willReturn(self::CUSTOM_KEY);

        $this->cachePoolMock
            ->expects(self::once())
            ->method('getItem')
            ->with(self::CUSTOM_KEY)
            ->willThrowException($cacheException);

        $result = $this->subject->check();

        self::assertFalse($result->isSuccess());
        self::assertEquals('custom error', $result->getMessage());
    }

    public function testForFailedCheckerResultWhenSavingItemFailed(): void
    {
        $cacheItemMock = $this->createMock(CacheItemInterface::class);

        $this->keyGeneratorMock
            ->expects(self::once())
            ->method('generate')
            ->willReturn(self::CUSTOM_KEY);

        $this->cachePoolMock
            ->expects(self::once())
            ->method('getItem')
            ->with(self::CUSTOM_KEY)
            ->willReturn($cacheItemMock);

        $this->cachePoolMock
            ->expects(self::once())
            ->method('save')
            ->with($cacheItemMock)
            ->willReturn(false);

        $result = $this->subject->check();

        self::assertFalse($result->isSuccess());
        self::assertEquals(sprintf('Writing item %s failed', self::CUSTOM_KEY), $result->getMessage());
    }

    public function testForFailedCheckerResultWhenItemHitMissed(): void
    {
        $cacheItemMock = $this->createMock(CacheItemInterface::class);
        $cacheItemMock
            ->expects(self::once())
            ->method('isHit')
            ->willReturn(false);

        $this->keyGeneratorMock
            ->expects(self::once())
            ->method('generate')
            ->willReturn(self::CUSTOM_KEY);

        $this->cachePoolMock
            ->expects(self::exactly(2))
            ->method('getItem')
            ->with(self::CUSTOM_KEY)
            ->willReturn($cacheItemMock);

        $this->cachePoolMock
            ->expects(self::once())
            ->method('save')
            ->with($cacheItemMock)
            ->willReturn(true);

        $result = $this->subject->check();

        self::assertFalse($result->isSuccess());
        self::assertEquals(sprintf('Missed hit on item %s', self::CUSTOM_KEY), $result->getMessage());
    }

    public function testForFailedCheckerResultWhenRetrievedValueIsNotTheSame(): void
    {
        $cacheItemMock = $this->createMock(CacheItemInterface::class);
        $cacheItemMock
            ->expects(self::once())
            ->method('isHit')
            ->willReturn(true);

        $cacheItemMock
            ->expects(self::once())
            ->method('get')
            ->willReturn('wrong-value');

        $this->keyGeneratorMock
            ->expects(self::once())
            ->method('generate')
            ->willReturn(self::CUSTOM_KEY);

        $this->cachePoolMock
            ->expects(self::exactly(2))
            ->method('getItem')
            ->with(self::CUSTOM_KEY)
            ->willReturn($cacheItemMock);

        $this->cachePoolMock
            ->expects(self::once())
            ->method('save')
            ->with($cacheItemMock)
            ->willReturn(true);

        $result = $this->subject->check();

        self::assertFalse($result->isSuccess());
        self::assertEquals(sprintf('Mismatched value on item %s', self::CUSTOM_KEY), $result->getMessage());
    }

    public function testForFailedCheckerResultWhenRemovingItemFailed(): void
    {
        $cacheItemMock = $this->createMock(CacheItemInterface::class);
        $cacheItemMock
            ->expects(self::once())
            ->method('isHit')
            ->willReturn(true);

        $cacheItemMock
            ->expects(self::once())
            ->method('get')
            ->willReturn(self::CUSTOM_KEY);

        $this->keyGeneratorMock
            ->expects(self::once())
            ->method('generate')
            ->willReturn(self::CUSTOM_KEY);

        $this->cachePoolMock
            ->expects(self::exactly(2))
            ->method('getItem')
            ->with(self::CUSTOM_KEY)
            ->willReturn($cacheItemMock);

        $this->cachePoolMock
            ->expects(self::once())
            ->method('save')
            ->with($cacheItemMock)
            ->willReturn(true);

        $this->cachePoolMock
            ->expects(self::once())
            ->method('deleteItem')
            ->with(self::CUSTOM_KEY)
            ->willReturn(false);

        $result = $this->subject->check();

        self::assertFalse($result->isSuccess());
        self::assertEquals(sprintf('Removing item %s failed', self::CUSTOM_KEY), $result->getMessage());
    }
}


tests/Unit/UuidCacheKeyGeneratorTest.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace OAT\Library\HealthCheckCache\Tests\Unit;

use OAT\Library\HealthCheckCache\UuidCacheKeyGenerator;
use PHPUnit\Framework\TestCase;
use Ramsey\Uuid\Uuid;
use Ramsey\Uuid\UuidFactory;
use Ramsey\Uuid\UuidInterface;

class UuidCacheKeyGeneratorTest extends TestCase
{
    private const TEST_UUID = '16dfb592-dbc2-4727-9fff-34db268fdddd';

    /**
     * @runInSeparateProcess
     * @preserveGlobalState disabled
     */
    public function testGenerateWithDefaultPrefix(): void
    {
        $this->prepareTestUuidFactory();

        $generator = new UuidCacheKeyGenerator();

        self::assertSame(sprintf('health-check-%s', self::TEST_UUID), $generator->generate());
    }

    /**
     * @runInSeparateProcess
     * @preserveGlobalState disabled
     */
    public function testGenerateWithCustomPrefix(): void
    {
        $this->prepareTestUuidFactory();

        $generator = new UuidCacheKeyGenerator('custom-prefix');

        self::assertSame(sprintf('custom-prefix-%s', self::TEST_UUID), $generator->generate());
    }

    private function prepareTestUuidFactory(): void
    {
        $factory = new class() extends UuidFactory
        {
            public $uuid;

            public function uuid4(): UuidInterface
            {
                return $this->uuid;
            }
        };
        Uuid::setFactory($factory);

        $factory->uuid = Uuid::fromString(self::TEST_UUID);
    }
}


tests/Unit/RamseyCacheKeyGeneratorTest.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace OAT\Library\HealthCheckCache\Tests\Unit;

use OAT\Library\HealthCheckCache\RamseyCacheKeyGenerator;
use PHPUnit\Framework\TestCase;
use Ramsey\Uuid\Uuid;
use Ramsey\Uuid\UuidFactory;
use Ramsey\Uuid\UuidInterface;

class RamseyCacheKeyGeneratorTest extends TestCase
{
    private const TEST_UUID = '16dfb592-dbc2-4727-9fff-34db268fdddd';

    /**
     * @runInSeparateProcess
     * @preserveGlobalState disabled
     */
    public function testGenerateWithDefaultPrefix(): void
    {
        $this->prepareTestUuidFactory();

        $generator = new RamseyCacheKeyGenerator();

        self::assertSame(sprintf('oat-health-check-%s', self::TEST_UUID), $generator->generate());
    }

    /**
     * @runInSeparateProcess
     * @preserveGlobalState disabled
     */
    public function testGenerateWithCustomPrefix(): void
    {
        $this->prepareTestUuidFactory();

        $generator = new RamseyCacheKeyGenerator('custom-prefix');

        self::assertSame(sprintf('custom-prefix-%s', self::TEST_UUID), $generator->generate());
    }

    private function prepareTestUuidFactory(): void
    {
        $factory = new class() extends UuidFactory {
            public $uuid;

            public function uuid4(): UuidInterface
            {
                return $this->uuid;
            }
        };
        Uuid::setFactory($factory);

        $factory->uuid = Uuid::fromString(self::TEST_UUID);
    }
}


src/UuidCacheKeyGenerator.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace OAT\Library\HealthCheckCache;

use Ramsey\Uuid\Uuid;

class UuidCacheKeyGenerator implements CacheKeyGeneratorInterface
{
    private const DEFAULT_PREFIX = 'health-check';

    /** @var string */
    private $prefix;

    public function __construct(string $prefix = self::DEFAULT_PREFIX)
    {
        $this->prefix = $prefix;
    }

    public function generate(): string
    {
        return sprintf('%s-%s', $this->prefix, Uuid::uuid4());
    }
}


src/RamseyCacheKeyGenerator.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace OAT\Library\HealthCheckCache;

use Ramsey\Uuid\Uuid;

final class RamseyCacheKeyGenerator implements CacheKeyGeneratorInterface
{
    private const DEFAULT_PREFIX = 'oat-health-check';

    /** @var string */
    private $prefix;

    public function __construct(string $prefix = null)
    {
        $this->prefix = $prefix ?? self::DEFAULT_PREFIX;
    }

    public function generate(): string
    {
        return sprintf('%s-%s', $this->prefix, Uuid::uuid4());
    }
}


src/CacheKeyGeneratorInterface.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace OAT\Library\HealthCheckCache;

interface CacheKeyGeneratorInterface
{
    public function generate(): string;
}


src/CacheChecker.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace OAT\Library\HealthCheckCache;

use OAT\Library\HealthCheck\Checker\CheckerInterface;
use OAT\Library\HealthCheck\Result\CheckerResult;
use Psr\Cache\CacheException;
use Psr\Cache\CacheItemPoolInterface;

class CacheChecker implements CheckerInterface
{
    private const IDENTIFIER = 'cache';

    /** @var CacheItemPoolInterface */
    private $cacheItemPool;

    /** @var CacheKeyGeneratorInterface */
    private $cacheKeyGenerator;

    public function __construct(CacheItemPoolInterface $cacheItemPool, CacheKeyGeneratorInterface $cacheKeyGenerator = null)
    {
        $this->cacheItemPool = $cacheItemPool;
        $this->cacheKeyGenerator = $cacheKeyGenerator ?? new UuidCacheKeyGenerator();
    }

    public function getIdentifier(): string
    {
        return self::IDENTIFIER;
    }

    public function check(): CheckerResult
    {
        try {
            $keyAndValue = $this->cacheKeyGenerator->generate();

            $item = $this->cacheItemPool->getItem($keyAndValue);
            $item->set($keyAndValue);
            $item->expiresAfter(30);

            if (!$this->cacheItemPool->save($item)) {
                return new CheckerResult(false, sprintf('Writing item %s failed', $keyAndValue));
            }

            $item = $this->cacheItemPool->getItem($keyAndValue);
            if (!$item->isHit()) {
                return new CheckerResult(false, sprintf('Missed hit on item %s', $keyAndValue));
            }

            if ($item->get() !== $keyAndValue) {
                return new CheckerResult(false, sprintf('Mismatched value on item %s', $keyAndValue));
            }

            if (!$this->cacheItemPool->deleteItem($keyAndValue)) {
                return new CheckerResult(false, sprintf('Removing item %s failed', $keyAndValue));
            }

            return new CheckerResult(
                true,
                sprintf('Success on writing, reading and deleting cache item %s', $keyAndValue)
            );
        } catch (CacheException $exception) {
            return new CheckerResult(false, $exception->getMessage());
        }
    }
}
